src/Payments.sol

```diff
@@ -393,45 +571,128 @@ contract Payments is
+                "terminated rail is beyond it's max settlement epoch; settle the rail"
+            );
+
+            require(
+                newRate <= oldRate,
+                "failed to modify rail: cannot increase rate on terminated rail"
+            );
+        }
+
+        if (lockupSettledUpto == block.number) {
+            // if account lockup is fully settled; there's nothing left to do
+            return;
+        }
+
+        // Case 2.A: Lockup not fully settled -> check if rail is in debt
+        if (block.number > payer.lockupLastSettledAt + rail.lockupPeriod) {
+            require(newRate == oldRate, "rail is in-debt; cannot change rate");
+            return;
+        }
+
+        // Case 2.B: Lockup not fully settled  but rail is not in debt -> check if rate is being increased
+        require(
+            newRate <= oldRate,
+            "account lockup not fully settled; cannot increase rate"
+        );
+    }
+
+    function updateOperatorLockupTracking(
+        OperatorApproval storage approval,
+        uint256 oldLockup,
+        uint256 newLockup
     ) internal {
-        // Ensure the one-time payment does not exceed the available fixed lockup on the rail.
+        if (newLockup < oldLockup) {
+            uint256 lockupDecrease = oldLockup - newLockup;
+            approval.lockupUsage -= lockupDecrease;
+            return;
+        }
+
+        // Handle lockup increase
+        uint256 lockupIncrease = newLockup - oldLockup;
+
+        // Verify against allowance
         require(
-            oneTimePayment <= rail.lockupFixed,
-            "one-time payment exceeds rail fixed lockup"
+            approval.lockupUsage + lockupIncrease <= approval.lockupAllowance,
+            "exceeds operator lockup allowance"
         );
 
-        // Calculate the original total lockup amount:
-        uint256 oldTotalLockup = (oldRate * rail.lockupPeriod) +
+        // Update usage
+        approval.lockupUsage += lockupIncrease;
+    }
+
+    function validateAndModifyRateChangeApproval(
+        Rail storage rail,
+        OperatorApproval storage approval,
+        uint256 oldRate,
+        uint256 newRate,
+        uint256 oneTimePayment,
+        uint256 effectiveLockupPeriod
+    ) internal {
+        // Handle rate-based lockup changes
+        uint256 oldLockup = (oldRate * effectiveLockupPeriod) +
             rail.lockupFixed;
-        uint256 newTotalLockup = (newRate * rail.lockupPeriod) +
+        uint256 newLockup = (newRate * effectiveLockupPeriod) +
             rail.lockupFixed;
 
-        require(
-            newTotalLockup <= oldTotalLockup + approval.lockupAllowance,
-            "exceeds operator lockup allowance"
+        // note: we've reduced `rail.lockupFixed` by `oneTimePayment` before calling this
+        // so we need to add the oneTimePayment to the new lockup for allowance tracking
+        // as all one time payments are also bound by the operator lockup allowance
```

Comment:
This looks correct, but it's really confusing.

Reply:
(not sure how to improve it, but we should consider revisiting it in a future PR)

Reply:
@Stebalien I agree. I am simplifying it in the current PR itself.

Reply:
@Stebalien Have simplified this and it looks much cleaner now. Please take a look.